### PR TITLE
macOS and Mac Catalyst support

### PR DESCRIPTION
**Pull request by mpodwysocki**
This is to add macOS and Mac Catalyst support for Azure Notification Hubs SDK

### CODE REVIEW

File: Scripts/build-archive.sh [hunk 1]
```diff
@@ -0,0 +1,127 @@
+#!/bin/bash
+
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
```
Comment (MatkovIvan) [inline Scripts/build-archive.sh:4]: I saw that you're using "Apache License 2.0" here

File: Scripts/build-archive.sh [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+# Get current version.
+VERSION="$($(dirname "$0")/framework-version.sh)"
+
+# Archive fat frameworks for CocoaPods.
+archive "$PRODUCT_NAME-$VERSION.zip" "$PRODUCT_NAME/iOS" "$PRODUCT_NAME/macOS" "$PRODUCT_NAME/tvOS"
+
+# Archive fat frameworks for Carthage.
+mv "$PRODUCTS_DIR/iOS/AppCenterDistributeResources.bundle" "$PRODUCTS_DIR/iOS/AppCenterDistribute.framework"
```
Comment (MatkovIvan) [inline Scripts/build-archive.sh:83]: There is no `AppCenterDistribute` here

File: Scripts/build-docs.sh [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+INSTALL_DIR="${PRODUCTS_DIR}/${PROJECT_NAME}.framework"
+DOCUMENTATION_DIR="${PRODUCTS_DIR}/Documentation/${PROJECT_NAME}"
+
+# Check if jazzy installed.
+if [ ! -x "$(command -v jazzy)" ]; then
+  echo "Couldn't find jazzy. Install jazzy before building frameworks"
+  exit 1
+fi
+
+jazzy --config "${SRCROOT}/../Documentation/${OS_NAME}/${PROJECT_NAME}/.jazzy.yaml"
```
Comment (MatkovIvan) [inline Scripts/build-docs.sh:21]: I do not see any `jazzy` configs in the repo, so I guess you do not need this script

File: Scripts/get-release-notes.sh [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+CHANGELOG_FILE="$PROJECT_DIR/CHANGELOG.md"
+
+change_log_found=false
+while IFS='' read -r line || -n "$line" ]]; do
+
+# If it is reading change log for the version
+if $change_log_found; then
+
+  # If it reads end of change log for the version
+  if [[ "$line" =~ "___" ]]; then
```
Comment (MatkovIvan) [inline Scripts/get-release-notes.sh:19]:
```suggestion
  if [[ "$line" == "___" ]]; then
```

File: Scripts/update-version.sh [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+# Update documentation version
+for file in `find $DOCUMENT_DIR -name '.jazzy.yaml' -type f`; do
+  sed -i '' 's/\(module_version: \).*/\1'$new_version'/g' $file
+done
+
+# Update CocoaPods version
+sed -i '' "s/\(\.version[[:space:]]*= \)\'.*\'$/\1'$new_version'/1" $PODSPEC_FILE
+
+# Update SwiftPM version
+sed -i '' 's/\(define("APP_CENTER_C_VERSION",[[:space:]]*to:*\).*/\1''"\\"'$new_version'\\""),''/g' $SWIFTPM_FILE
```
Comment (MatkovIvan) [inline Scripts/update-version.sh:49]: You don't have `APP_CENTER_C_VERSION` define

File: Scripts/update-version.sh [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+SWIFTPM_FILE="$PROJECT_DIR/Package.swift"
+
+# Update framework version
+$(dirname "$0")/framework-version.sh $new_version
+
+# Update documentation version
+for file in `find $DOCUMENT_DIR -name '.jazzy.yaml' -type f`; do
+  sed -i '' 's/\(module_version: \).*/\1'$new_version'/g' $file
+done
+
```
Comment (MatkovIvan) [inline Scripts/update-version.sh:44]: There are no jazzy configs

File: WindowsAzureMessaging/WindowsAzureMessaging/Helpers/SBNotificationHubHelper.m [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+
+static const char encodingTable[] = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
+static char decodingTable[128];
+static NSString *decodingTableLock = @"decodingTableLock";
+static NSString *const domain = @"WindowsAzureMessaging";
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
++ (NSString *)urlEncode:(NSString *)urlString {
+    return (__bridge NSString *)CFURLCreateStringByAddingPercentEscapes(kCFAllocatorDefault, (__bridge CFStringRef)urlString, NULL,
```
Comment (MatkovIvan) [inline WindowsAzureMessaging/WindowsAzureMessaging/Helpers/SBNotificationHubHelper.m:20]: It seems that it should be `__bridge_transfer` to transfer ownership and avoid leak

File: WindowsAzureMessaging/WindowsAzureMessaging/HttpClient/Util/MSHttpUtil.h [hunk 8]
```diff
@@ -0,0 +1,48 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
```
Comment (MatkovIvan) [inline WindowsAzureMessaging/WindowsAzureMessaging/HttpClient/Util/MSHttpUtil.h:2]: MIT?

File: WindowsAzureMessaging/WindowsAzureMessaging/HttpClient/Util/MSHttpUtil.h [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+#import <Foundation/Foundation.h>
+
+static short const kMSMaxCharactersDisplayedForAppSecret = 8;
+static NSString *const kMSHidingStringForAppSecret = @"*";
```
Comment (MatkovIvan) [inline WindowsAzureMessaging/WindowsAzureMessaging/HttpClient/Util/MSHttpUtil.h:7]: Not sure that you need app_secret stuff

File: WindowsAzureMessaging/WindowsAzureMessaging/Support/WindowsAzureMessaging.xcconfig [hunk 10]
```diff
@@ -0,0 +1,5 @@
+PRODUCT_BUNDLE_IDENTIFIER = com.microsoft.WindowsAzureMessaging
+
+APPLICATION_EXTENSION_API_ONLY = NO;
```
Comment (MatkovIvan) [inline WindowsAzureMessaging/WindowsAzureMessaging/Support/WindowsAzureMessaging.xcconfig:3]:
```suggestion
APPLICATION_EXTENSION_API_ONLY = NO
```

File: WindowsAzureMessaging/WindowsAzureMessaging/Support/iOS WindowsAzureMessaging.xcconfig [hunk 11]
```diff
@@ -0,0 +1,4 @@
+#include "../Config/iOS.xcconfig"
+#include "../Config/Framework.xcconfig"
+
+OTHER_LDFLAGS=$(inherited) -framework UIKit  -weak_framework UserNotifications
```
Comment (MatkovIvan) [inline WindowsAzureMessaging/WindowsAzureMessaging/Support/iOS WindowsAzureMessaging.xcconfig:4]:
```suggestion
OTHER_LDFLAGS=$(inherited) -framework UIKit -weak_framework UserNotifications
```

File: Scripts/build-archive.sh (hunk 1 above)
Comment (mpodwysocki) [reply Scripts/build-archive.sh:4]: Removed license per requirements

File: Scripts/build-docs.sh (hunk 3 above)
Comment (mpodwysocki) [reply Scripts/build-docs.sh:21]: Yeah, planning on eventually adding docs but that's a longer term goal

File: Scripts/build-archive.sh (hunk 2 above)
Comment (mpodwysocki) [reply Scripts/build-archive.sh]: Removed

File: Scripts/get-release-notes.sh (hunk 4 above)
Comment (mpodwysocki) [reply Scripts/get-release-notes.sh]: Fixed

File: WindowsAzureMessaging/WindowsAzureMessaging/Helpers/SBNotificationHubHelper.m (hunk 7 above)
Comment (mpodwysocki) [reply WindowsAzureMessaging/WindowsAzureMessaging/Helpers/SBNotificationHubHelper.m]: Fixed

File: WindowsAzureMessaging/WindowsAzureMessaging/HttpClient/Util/MSHttpUtil.h (hunk 8 above)
Comment (mpodwysocki) [reply WindowsAzureMessaging/WindowsAzureMessaging/HttpClient/Util/MSHttpUtil.h]: Fixed all headers

File: WindowsAzureMessaging/WindowsAzureMessaging/HttpClient/Util/MSHttpUtil.h (hunk 9 above)
Comment (mpodwysocki) [reply WindowsAzureMessaging/WindowsAzureMessaging/HttpClient/Util/MSHttpUtil.h]: Removed

File: Scripts/update-version.sh (hunk 5 above)
Comment (mpodwysocki) [reply Scripts/update-version.sh]: Fixed to `NH_C_VERSION`

File: Package.swift [hunk 12]
```diff
@@ -1,26 +1,41 @@
...
             targets: ["WindowsAzureMessaging"]),
     ],
     dependencies: [],
     targets: [
         .target(
             name: "WindowsAzureMessaging",
-            path: "src/WindowsAzureMessaging/WindowsAzureMessaging",
-            sources: sources,
-            cSettings: sources.map { CSetting.headerSearchPath($0) }
+            path: "WindowsAzureMessaging/WindowsAzureMessaging",
```
Comment (MatkovIvan) [inline Package.swift:25]: AppCenter SDK has [inclide/*.h](https://github.com/microsoft/appcenter-sdk-apple/tree/3.3.0/AppCenter/AppCenter/include) links to public headers. Does this work correctly without similar thing?

File: Samples/Objective-C/nhubsample-refresh/nhubsample-refresh.xcodeproj/project.pbxproj [hunk 13]
```diff
@@ -359,6 +386,7 @@
 				);
 				PRODUCT_BUNDLE_IDENTIFIER = "com.example.nhubsample-refresh";
 				PRODUCT_NAME = "$(TARGET_NAME)";
+				SUPPORTS_MACCATALYST = YES;
 				TARGETED_DEVICE_FAMILY = "1,2";
```
Comment (MatkovIvan) [inline Samples/Objective-C/nhubsample-refresh/nhubsample-refresh.xcodeproj/project.pbxproj:390]: Additional settings like this are checked by `VerifyNoBuildSettings.swift` to force move everything like that to config files.  Probably something goes wrong in this check if it compiles without error.

File: Package.swift (hunk 12 above)
Comment (mpodwysocki) [reply Package.swift:25]: No, I'll have to do that for Swift Package Manager to work

File: Samples/Objective-C/nhubsample-refresh/nhubsample-refresh.xcodeproj/project.pbxproj (hunk 13 above)
Comment (mpodwysocki) [reply Samples/Objective-C/nhubsample-refresh/nhubsample-refresh.xcodeproj/project.pbxproj:390]: Right, I need to move all things from the sample apps both to their own support files

**Pull request merged**